reSnap: init at 2.5.2

404Wolf opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4426

eljamm commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,31 @@
+    hash = "sha256-P+q5NMGtXE2XCcGGHZJGw8RoOqJSHSv7oyMLkn/ltfY=";
+  }}/reSnap.sh";
+
+  meta = with lib; {
+    description = " Take screnshots of your reMarkable tablet over SSH";
+    homepage = "https://github.com/cloudsftp/reSnap";
+    license = with licenses; [mit];
+    maintainers = with maintainers; [_404wolf];
+    mainProgram = "reSnap";
+  };
```
```suggestion
  meta = {
    description = " Take screnshots of your reMarkable tablet over SSH";
    homepage = "https://github.com/cloudsftp/reSnap";
    license = with lib.licenses; [ mit ];
    maintainers = with lib.maintainers; [ _404wolf ];
    mainProgram = "reSnap";
  };
```

eljamm commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,31 @@
+writeShellApplication {
+  name = "reSnap";
+
+  runtimeInputs = with pkgs; [
+    ffmpeg
+    feh
+    imagemagick_light
+    lz4
+  ];
```
Don't use `pkgs` inside the derivation. Instead, add the dependencies to the inputs.

```suggestion
  runtimeInputs = [
    ffmpeg
    feh
    imagemagick_light
    lz4
  ];
```

eljamm commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,31 @@
+{
+  lib,
+  pkgs,
+  fetchFromGitHub,
+  writeShellApplication,
```
Don't use `pkgs` inside the derivation. Instead, add the dependencies to the inputs.

```suggestion
  lib,
  fetchFromGitHub,
  writeShellApplication,
  ffmpeg,
  feh,
  imagemagick_light,
  lz4,
```

eljamm commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,31 @@
+    lz4
+  ];
+
+  text = "${fetchFromGitHub {
```
File should be formatted with [nixfmt](https://github.com/NixOS/nixfmt)

eljamm commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,31 @@
+  fetchFromGitHub,
+  writeShellApplication,
+}:
+writeShellApplication {
```
Since we're using `writeShellApplication`, we don't have a `version` or `src` attributes, but they're pretty useful to have, so we can do something like this:

```suggestion
let
  name = "reSnap";
  version = "2.5.2";

  src = fetchFromGitHub {
    owner = "cloudsftp";
    repo = "reSnap";
    rev = "v${version}";
    hash = "sha256-P+q5NMGtXE2XCcGGHZJGw8RoOqJSHSv7oyMLkn/ltfY=";
  };
in
writeShellApplication {
  inherit name;

  runtimeInputs = [
    feh
    ffmpeg
    imagemagick_light
    lz4
  ];

  text = "${src}/reSnap.sh";

  meta = {
    description = " Take screnshots of your reMarkable tablet over SSH";
    homepage = "https://github.com/cloudsftp/reSnap";
    license = with lib.licenses; [ mit ];
    maintainers = with lib.maintainers; [ _404wolf ];
    mainProgram = "reSnap";
  };
}
```

Also, the latest git revision seems to be the same as the latest tagged release (2.5.2), so we should use that, instead. 

Can you update the commit message and the PR title to reflect this?

Commit:
reSnap: init at v2.5.2

404Wolf commented:
Cool, I've made all the changes, but I rebased the wrong things and am fixing a git issue now. This should be ready to merge soon

404Wolf commented:
I think it's all fixed and ready! Thanks again for all the great feedback!

eljamm commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,48 @@
+  name = "reSnap";
+  version = "2.5.2";
+
+  src = stdenv.mkDerivation {
```
I see that you're using `mkDerivation` to patch the script name. However, we should use `stdenvNoCC` since we're not compiling anything.

```suggestion
  src = stdenvNoCC.mkDerivation {
```

404Wolf replied:
Oh, super cool, thanks for showing me this!

eljamm commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,48 @@
+      repo = "reSnap";
+      rev = "v${version}";
+      hash = "sha256-thVyf1gTDPLQVtZKoWL7SGiWI++ICWqmF/Ar57I3WP8=";
+    };
```
Since we're just patching and copying a shell script, we only need these phases:

```suggestion
    };
    phases = [
      "unpackPhase"
      "patchPhase"
      "installPhase"
    ];
```

SuperSandro2000 replied:
Setting phases in packages is strictly forbidden. Also if the phases do not cause harm, just leave them enabled.

eljamm replied:
> Setting phases in packages is strictly forbidden

@SuperSandro2000 According to the [stdenv docs](https://github.com/NixOS/nixpkgs/blob/master/doc/stdenv/stdenv.chapter.md#phases-var-stdenv-phases), it's not forbidden to set this. It's just discouraged.

> It is discouraged to set this variable, as it is easy to miss some important functionality hidden in some of the less obviously needed phases (like `fixupPhase` which patches the shebang of scripts).
Usually, if you just want to add a few phases, it’s more convenient to set one of the variables below (such as `preInstallPhases`).

But perhaps it's that worth it after all.

eljamm reviewed (approved):
LGTM!

SuperSandro2000 commented on maintainers/maintainer-list.nix:
Please rebase

SuperSandro2000 commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,53 @@
+      hash = "sha256-thVyf1gTDPLQVtZKoWL7SGiWI++ICWqmF/Ar57I3WP8=";
+    };
+    installPhase = ''cp ./reSnap.sh $out'';
+    phases = [
+      "unpackPhase"
+      "patchPhase"
+      "installPhase"
+    ];
```
```suggestion
```

SuperSandro2000 commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,53 @@
+    '';
+  };
+in
+writeShellApplication {
```
Why not copy the code from src to a subdirectory in out and have a proper application with pname, version, etc?

SuperSandro2000 commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,53 @@
+      "installPhase"
+    ];
+    postPatch = ''
+      substituteInPlace ./reSnap.sh --replace-quiet "\$0" "reSnap"
```
```suggestion
      substituteInPlace ./reSnap.sh --replace-fail "\$0" "reSnap"
```

SuperSandro2000 commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,53 @@
+  fetchFromGitHub,
+}:
+let
+  name = "reSnap";
```
```suggestion
  pname = "resnap";
```

eljamm replied:
writeShellApplication has no `pname` attribute. You'd want to do this with `stdenv.mkDerivation`

404Wolf commented:
Hey @SuperSandro2000, good call about wrapping with mkDerivation, I've made the change. If I'm reading right you want to remove explicit phases -- why is this? And how does the current state look? Thanks!

404Wolf commented:
@SuperSandro2000 not to be annoying, but have you had a chance to look this over again?

eljamm commented:
Hey @404Wolf, I think you're doing great so far, but I believe the package can still be improved further:

- We don't need two `mkDerivation` since we can put everything in one
- We don't need `writeShellApplication` anymore as we can just copy the script to the output
- Use `finalAttrs` instead of `rec` (see <https://github.com/NixOS/nixpkgs/issues/315337>).
- Whenever we override a phase like `installPhase`, we should always add the `pre-` and `post-` hooks.
- Contrary to my first suggestion, it's discouraged to set `phases` according to the [stdenv docs](https://github.com/NixOS/nixpkgs/blob/420d0cc6f27b2b7f117be67aa721dd7154ffb003/doc/stdenv/stdenv.chapter.md#phases-var-stdenv-phases) since it's easy to forget those that run important things.

As such, here is what I think the derivation should look like:

<details><summary>package.nix</summary>

```nix
{
  lib,
  stdenvNoCC,
  fetchFromGitHub,
  ffmpeg,
  feh,
  imagemagick_light,
  lz4,
}:
stdenvNoCC.mkDerivation (finalAttrs: {
  pname = "resnap";
  version = "2.5.2";

  src = fetchFromGitHub {
    owner = "cloudsftp";
    repo = "reSnap";
    rev = "v${finalAttrs.version}";
    hash = "sha256-thVyf1gTDPLQVtZKoWL7SGiWI++ICWqmF/Ar57I3WP8=";
  };

  buildInputs = [
    ffmpeg
    feh
    imagemagick_light
    lz4
  ];

  installPhase = ''
    runHook preInstall

    install -D ./reSnap.sh $out/bin/reSnap

    runHook postInstall
  '';

  postFixup = ''
    substituteInPlace $out/bin/reSnap \
      --replace-fail "\$0" "reSnap"
  '';

  meta = {
    description = "Take screnshots of your reMarkable tablet over SSH";
    homepage = "https://github.com/cloudsftp/reSnap";
    license = with lib.licenses; [ mit ];
    maintainers = with lib.maintainers; [ _404wolf ];
    mainProgram = "reSnap";
  };
})
```

</details>

Let me know what you think and if you have any questions regarding this, don't hesitate to ask.

404Wolf commented:
> Hey @404Wolf, I think you're doing great so far, but I believe the package can still be improved further:
> 
>     * We don't need two `mkDerivation` since we can put everything in one
> 
>     * We don't need `writeShellApplication` anymore as we can just copy the script to the output
> 
>     * Use `finalAttrs` instead of `rec` (see [Documentation: guide for using `let in` vs `rec` vs `finalAttrs` #315337](https://github.com/NixOS/nixpkgs/issues/315337)).
> 
>     * Whenever we override a phase like `installPhase`, we should always add the `pre-` and `post-` hooks.
> 
>     * Contrary to my first suggestion, it's discouraged to set `phases` according to the [stdenv docs](https://github.com/NixOS/nixpkgs/blob/420d0cc6f27b2b7f117be67aa721dd7154ffb003/doc/stdenv/stdenv.chapter.md#phases-var-stdenv-phases) since it's easy to forget those that run important things.
> 
> 
> As such, here is what I think the derivation should look like:
> package.nix
> 
> Let me know what you think and if you have any questions regarding this, don't hesitate to ask.

Thanks, I've changed it to be that, that makes a ton of sense. I didn't realize that `buildInputs` also propagate to runtime. 

> Hey @404Wolf, I think you're doing great so far, but I believe the package can still be improved further:
> 
>     * We don't need two `mkDerivation` since we can put everything in one
> 
>     * We don't need `writeShellApplication` anymore as we can just copy the script to the output
> 
>     * Use `finalAttrs` instead of `rec` (see [Documentation: guide for using `let in` vs `rec` vs `finalAttrs` #315337](https://github.com/NixOS/nixpkgs/issues/315337)).
> 
>     * Whenever we override a phase like `installPhase`, we should always add the `pre-` and `post-` hooks.
> 
>     * Contrary to my first suggestion, it's discouraged to set `phases` according to the [stdenv docs](https://github.com/NixOS/nixpkgs/blob/420d0cc6f27b2b7f117be67aa721dd7154ffb003/doc/stdenv/stdenv.chapter.md#phases-var-stdenv-phases) since it's easy to forget those that run important things.
> 
> 
> As such, here is what I think the derivation should look like:
> package.nix
> 
> Let me know what you think and if you have any questions regarding this, don't hesitate to ask.

I could be wrong, but I don't think that buildInputs are available at runtime; it didn't seem to work. I cleaned it up a bit by only using one `mkDerivation` though, and just manually append to the PATH instead of writeShellApplication, and by reusing the name. Let me know what you think. Thanks for looking it over :)

eljamm commented:
I believe `runtimeInputs` will work in this case, but if that doesn't, we can wrap the executable with the PATH:

```nix
  nativeBuildInputs = [ makeWrapper ];

  runtimeInputs = [
    ffmpeg
    feh
    imagemagick_light
    lz4
  ];

  postFixup = ''
    substituteInPlace $out/bin/reSnap \
      --replace-fail "\$0" "reSnap"

    wrapProgram $out/bin/reSnap \
      --set PATH "${lib.makeBinPath finalAttrs.runtimeInputs}"
  '';
```

eljamm commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,54 @@
+    homepage = "https://github.com/cloudsftp/reSnap";
+    license = with lib.licenses; [ mit ];
+    maintainers = with lib.maintainers; [ _404wolf ];
+    mainProgram = name;
```
Changes to `meta` should not cause a rebuild, which would happen if `name` changes. We should explicitly set this.

```suggestion
    mainProgram = "reSnap";
```

eljamm commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,54 @@
+      --replace-fail "\$0" ${name}
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/bin
+    echo PATH="\$PATH:${
+      lib.strings.makeBinPath [
+        ffmpeg
+        feh
+        imagemagick_light
+        lz4
+      ]
+    }" > $out/bin/reSnap
+    cat ./reSnap.sh >> $out/bin/reSnap
+    chmod +x $out/bin/reSnap
+
+    runHook postInstall
+  '';
```
We can use `runtimeInputs` or wrap the executable instead of doing this.

```suggestion
   installPhase = ''
    runHook preInstall

    install -D ./reSnap.sh $out/bin/reSnap

    runHook postInstall
  '';
```

SuperSandro2000 commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,54 @@
+
+  postFixup = ''
+    substituteInPlace $out/bin/reSnap \
+      --replace-fail "\$0" ${name}
```
```suggestion
      --replace-fail "\$0" reSnap
```

SuperSandro2000 commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,54 @@
+  name = "reSnap";
+in
+stdenvNoCC.mkDerivation (finalAttrs: {
+  pname = lib.strings.toLower name;
+  version = "2.5.2";
+
+  src = fetchFromGitHub {
+    owner = "cloudsftp";
+    repo = name;
```
```suggestion
  pname = "resnap;
  version = "2.5.2";

  src = fetchFromGitHub {
    owner = "cloudsftp";
    repo = "reSnap";
```

SuperSandro2000 commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,54 @@
+  imagemagick_light,
+  lz4,
+}:
+let
+  name = "reSnap";
+in
```
```suggestion

```

eljamm reviewed (approved):
LGTM

404Wolf commented:
Hey @SuperSandro2000, any chance you could look this over at some point? I think it's ready to merge. Thanks!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4607

lelgenio commented on pkgs/by-name/re/resnap/package.nix:
```diff
@@ -0,0 +1,54 @@
+    lz4
+  ];
+
+  postFixup = ''
+    substituteInPlace $out/bin/reSnap \
+      --replace-fail "\$0" reSnap
+
+    wrapProgram $out/bin/reSnap \
+      --suffix PATH : "${lib.makeBinPath finalAttrs.runtimeInputs}"
+  '';
+
+  installPhase = ''
+    runHook preInstall
+
+    install -D reSnap.sh $out/bin/reSnap
+
+    runHook postInstall
+  '';
```
```suggestion
  installPhase = ''
    runHook preInstall

    install -D reSnap.sh $out/bin/reSnap

    runHook postInstall
  '';

  postFixup = ''
    substituteInPlace $out/bin/reSnap \
      --replace-fail "\$0" reSnap

    wrapProgram $out/bin/reSnap \
      --suffix PATH : "${lib.makeBinPath finalAttrs.runtimeInputs}"
  '';
```

Fixup is executed after install

wolfgangwalther committed:
Update pkgs/by-name/re/resnap/package.nix

Co-authored-by: Leonardo Eugênio <lelgenio@lelgenio.com>

404Wolf commented:
@wolfgangwalther thanks! Sorry for not following up myself, I'm excited to see this get merged.
